Fix a bug of `FTPArtifactRepository.log_artifacts` with `artifact_path` keyword argument (issue #3388)

kzm4269 opened the pull request:
## What changes are proposed in this pull request?

- Fixes #3388
- Adds a test case for `FTPArtifactRepository.log_artifacts`.

## How is this patch tested?

Verify no exisiting tests in `tests/store/artifact/test_ftp_artifact_repo.py` are broken.

## Release Notes

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

kzm4269 committed:
Test the full path of the logged artifacts

The test for `FTPArtifactRepository.log_artifacts` must check the full
path of the logged artifacts.

Signed-off-by: Kazuma Takahara <4269kzm@gmail.com>

kzm4269 committed:
Cover a test case where a directory to log contains directories

Signed-off-by: Kazuma Takahara <4269kzm@gmail.com>

kzm4269 committed:
Fix a bug of `FTPArtifactRepository.log_artifacts` (#3388)

Signed-off-by: Kazuma Takahara <4269kzm@gmail.com>

dbczumar reviewed (approved):
@kzm4269 LGTM! Thanks for this fix! Can you address the linter and DCO check errors in order to merge this PR? Here are the instructions for passing the DCO check by signing your commits: https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#sign-your-work

harupy commented on tests/store/artifact/test_ftp_artifact_repo.py:
```diff
@@ -211,32 +211,99 @@ def test_log_artifact_multiple_calls(ftp_mock, tmpdir):
 
+def __posixpath_parents(pathname, root):
+    parents = []
+    parent, _ = posixpath.split(pathname)
```
Can we use `posixpath.dirname` here?

harupy commented on tests/store/artifact/test_ftp_artifact_repo.py:
```diff
@@ -211,32 +211,99 @@ def test_log_artifact_multiple_calls(ftp_mock, tmpdir):
+    dest_path = (
+        dest_path_root if artifact_path is None else posixpath.join(dest_path_root, artifact_path)
+    )
+    dir_expected = set(
```
```suggestion
    dirs_expected = set(
```

to indicate this variable represents a collection of directories?

harupy commented on tests/store/artifact/test_ftp_artifact_repo.py:
```diff
@@ -211,32 +211,99 @@ def test_log_artifact_multiple_calls(ftp_mock, tmpdir):
+    )
 
-    ftp_mock.cwd = MagicMock(side_effect=[ftplib.error_perm, None, None, None, None, None])
+    for dir_expected_i in set(dir_expected):
```
```suggestion
    for dir_expected_i in dir_expected:
```

because `dir_excepted` is already set? Please let me know if I'm missing something.

harupy commented on tests/store/artifact/test_ftp_artifact_repo.py:
```diff
@@ -211,32 +211,99 @@ def test_log_artifact_multiple_calls(ftp_mock, tmpdir):
+        parent, _ = posixpath.split(pathname)
+        if parent not in dir_created:
+            raise ftplib.error_perm
+        assert pathname not in dir_created, "mock not implemented"
```
Do we really need this assrertion (`assert pathname not in dir_created`)?

harupy commented on tests/store/artifact/test_ftp_artifact_repo.py:
```diff
@@ -211,32 +211,99 @@ def test_log_artifact_multiple_calls(ftp_mock, tmpdir):
+    cwd_history = []
+
+    def mkd_mock(pathname):
+        assert pathname == posixpath.abspath(pathname), "mock not implemented"
```
`"mock not implemented"` doesn't seem to be an appropriate message for the assert condition `pathname == posixpath.abspath(pathname)`.

kzm4269 committed:
Revise in response to review comments

- Use `posixpath.dirname` instead of `posixpath.split` to get a parent directory path.
- Rename collection variables.
- Remove a duplicated code.
- Implement `mkd_mock` and `cwd_mock` for a relative `pathname` argment.
- Indicates that the set is being copied for in-place processing by using `set.copy()` instead of `set()`.

Signed-off-by: Kazuma Takahara <4269kzm@gmail.com>
Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

harupy commented on tests/store/artifact/test_ftp_artifact_repo.py:
```diff
@@ -211,32 +211,93 @@ def test_log_artifact_multiple_calls(ftp_mock, tmpdir):
     assert ftp_mock.storbinary.call_args_list[0][0][0] == "STOR test2.txt"
 
 
+def __posixpath_parents(pathname, root):
+    parents = [posixpath.dirname(pathname)]
+    root = posixpath.normpath(root)
+    while parents[-1] != "/" and parents[-1] != root:
+        parents.append(posixpath.dirname(parents[-1]))
+    return parents
```
correct me if I'm wrong, but I think this function assumes:

- `pathname` is an abosolute path
-  `pathname` startswith `root`

Can we add verify these conditions with `assert` to prevent this function from falling into an infinite loop?

harupy replied:
```suggestion
def __posixpath_parents(pathname, root):
    assert posixpath.isabs(pathname)
    assert pathname.startswith(root)
    
    parents = [posixpath.dirname(pathname)]
    root = posixpath.normpath(root)
    while parents[-1] != "/" and parents[-1] != root:
        parents.append(posixpath.dirname(parents[-1]))
    return parents
```

harupy commented:
Wrote a simple script to verify the fix works properly:


### docker-compose.yml:

```yml
version: "3"
services:
  ftp:
    image: stilliard/pure-ftpd
    environment:
      FTP_USER_NAME: mlflow
      FTP_USER_PASS: mlflow
      FTP_USER_HOME: /home/mlflow

    ports:
      - "21:21"
      - "30000-30009:30000-30009"

```

### ftp_artifacts.py:

```python
import os
import uuid
import tempfile
from pathlib import Path

import mlflow


tmpdir = tempfile.mkdtemp()

# setup a temporary experiment
tracking_uri = "file://{}".format(os.path.join(tmpdir, "mlruns"))
mlflow.set_tracking_uri(tracking_uri)

expr_name = uuid.uuid4().hex
artifact_location = "ftp://mlflow:mlflow@localhost:21/mlflow"
mlflow.create_experiment(expr_name, artifact_location=artifact_location)
mlflow.set_experiment(expr_name)

# create files to log
p = Path(tmpdir).joinpath("files")
p.mkdir()
p.joinpath("a.txt").write_text("a")
p.joinpath("b.txt").write_text("b")
p.joinpath("subdir").mkdir()
p.joinpath("subdir").joinpath("c.txt").write_text("c")
p.joinpath("subdir").joinpath("d.txt").write_text("d")
p.joinpath("empty").mkdir()

# log files with `artifact_path`
with mlflow.start_run() as run:
    mlflow.log_artifacts(p.absolute(), artifact_path="subdir")


def yield_artifacts(run_id, path=None):
    client = mlflow.tracking.MlflowClient()
    for item in client.list_artifacts(run_id, path):
        if item.is_dir:
            yield item.path
            yield from yield_artifacts(run_id, item.path)
        else:
            yield item.path


# show artifacts
for a in yield_artifacts(run.info.run_id):
    print(a)
```

### how to run:

```
docker-compose up
python ftp_artifacts.py
```

### otuput:

before:

```
empty
subdir
subdir/a.txt
subdir/b.txt
subdir/c.txt
subdir/d.txt
```

after:

```
subdir
subdir/a.txt
subdir/b.txt
subdir/empty
subdir/subdir
subdir/subdir/c.txt
subdir/subdir/d.txt
```

kkendzia commented:
Do you know when this will be merged? Really looking forward for this fix.

kkendzia commented:
@kzm4269 Do you have to change anything so they can merge it?
